flang/lib/Optimizer/Dialect/FIROps.cpp

```diff
@@ -991,6 +991,23 @@ void fir::InsertOnRangeOp::build(mlir::OpBuilder &builder,
   build(builder, result, resTy, aggVal, eleVal, aa);
 }
 
+/// Range bounds must be nonnegative, and the range must not be empty.
+static mlir::LogicalResult verify(fir::InsertOnRangeOp op) {
+  bool rangeIsKnownToBeNonempty = false;
+  for (auto i = op.coor().end(), b = op.coor().begin(); i != b;) {
+    int64_t ub = (*--i).cast<IntegerAttr>().getInt();
+    int64_t lb = (*--i).cast<IntegerAttr>().getInt();
+    if (lb < 0 || ub < 0)
+      return op.emitOpError("negative range bound");
+    if (rangeIsKnownToBeNonempty)
```

Comment:
I do not follow the logic with `rangeIsKnownToBeNonempty`, why is not an issue if any of the bounds are such that `lb > ub`, if any of the range extents is empty, then isn't the range empty ?

Reply:
Not sure I follow this either. Just because a column has a well-defined range doesn't really mean that an invalid sized row (with a non-existent range) is valid. To insert the value into the array value, a full and complete set of access coordinates is required, right? Otherwise there is nowhere to put the value, meaning there is no element to insert the value into, meaning the Op is invalid semantically.

Reply:
There may be confusion about what this code is checking.  It is **not** checking if a particular `InsertOnRangeOp` bound value is in the correct range for a dimension.  The op doesn't supply that information, that cannot be checked.  This op is **not** a Fortran slice.  It describes the end points of an subsequence of the sequence of array elements in array element order (9.5.3.2).

Reply:
Here is an example.  The `fir.insert_on_range` op defining `%2` has `(lb2=1, lb1=2) <= (ub2=2, ub1=1)` which is valid, although `lb1 > ub1`.  Neither `fir.insert_on_range` op is a rectangular slice.
```
  integer, dimension(3,3) :: rr
  rr = reshape([5, 5, 5, 5, 5, 7, 7, 7, 9], shape(rr))
```

```
fir.global internal @_QQro.3x3xi4.b5ebfe96c7c5b7021237cf0045c79728 constant : !fir.array<3x3xi32> {
  %c5_i32 = constant 5 : i32
  %c7_i32 = constant 7 : i32
  %c9_i32 = constant 9 : i32
  %0 = fir.undefined !fir.array<3x3xi32>

  %1 = %0, %c5_i32, [0 : index, 1 : index, 0 : index, 1 : index] : (!fir.array<3x3xi32>, i32) -> !fir.array<3x3xi32>
           [0,0] [1,0] [2,0] [0,1] [1,1]

  %2 = fir.insert_on_range %1, %c7_i32, [2 : index, 1 : index, 1 : index, 2 : index] : (!fir.array<3x3xi32>, i32) -> !fir.array<3x3xi32>
           [2,1] [0,2] [1,2]

  %3 = fir.insert_value %2, %c9_i32, [2 : index, 2 : index] : (!fir.array<3x3xi32>, i32) -> !fir.array<3x3xi32>
           [2,2]

  fir.has_value %3 : !fir.array<3x3xi32>
}
```

Reply:
Ok, I was indeed confused and expected a Fortran stride while this really is an operation that inserts values between coordinate_1 and coordinate_2 in Fortran sequence order. Thanks for the detailed examples and explanations. Then it all makes sense to me.

Reply:
Thanks for the explanation, Val.